Optimize Value Suggestions in Autocomplete

LDrago27 opened the pull request:
### Description

<!-- Describe what this change achieves-->
Currently for larger datasets the process of fetching column Values is quite slow , making the autocomplete experience quite laggy. This PR introduces the following in a bid to reduce the wait time. 

- We introduce 2 new fields to IndexPattern **autoCompleteValues** and **topAggValues**

Whenever a value suggestion is triggered the following flow is executed
- If we already have results cached in autoCompleteValues we show those values
- If we don't have results cached, we try to showcase the  **Top Five Values** computed by sidebar to user and fire an async query to fetch column values from the API and update the cache in background
- In case we don't have **Top Five Values**, we synchronously fire the query to fetch column values and update the cache

Comparison:
Before:
https://github.com/user-attachments/assets/e69f6e5d-57c3-43c3-93bd-a5b9c8f29229

We were triggering the API call to fetch values each time, and on average in the above scenario was taking around 6s to show values on screen. (Anecdotal evidence based on the testing on above dataset)

After the change:
https://github.com/user-attachments/assets/1a94a59c-888c-4a02-aa5e-f7c6c2f85d2c

Now if we have the value present in the current query results we show the top values from the query result, while updating the cache using the API call. For all subsequent calls we are using the cached values from API call. 
So we are now showing suggestions to user instantaneously, also the API call durations have been slashed to around 1.5s. (Anecdotal evidence based on the testing on above dataset)

### Issues Resolved

<!-- List any issues this PR will resolve. Prefix the issue with the keyword closes, fixes, fix -->
<!-- Example: closes #1234 or fixes <Issue_URL> -->

## Screenshot

<!-- Attach any relevant screenshots. Any change to the UI requires an attached screenshot in the PR Description -->

## Testing the changes

<!--
  Please provide detailed steps for validating your changes. This could involve specific commands to run,
  pages to visit, scenarios to try or any other information that would help reviewers verify
  the functionality of your change
-->

## Changelog
<!--
Add a short but concise sentence about the impact of this pull request. Prefix an entry with the type of change they correspond to: breaking, chore, deprecate, doc, feat, fix, infra, refactor, test.
- fix: Update the graph
- feat: Add a new feature

If this change does not need to added to the changelog, just add a single `skip` line e.g.
- skip

Descriptions following the prefixes must be 100 characters long or less
-->
- fix: Optimize Value Suggestions in Autocomplete 
### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

kavilla commented on src/plugins/data/common/index_patterns/types.ts:
```diff
@@ -185,6 +185,8 @@ export interface FieldSpec {
   readFromDocValues?: boolean;
   subType?: IFieldSubType;
   indexed?: boolean;
+  autoCompleteValues?: string[];
```
Could we use the data views field spec instead ? That way we know we aren't messing with non experimental stuff

kavilla replied:
I don't think "auto complete" would be the sole consumer of this. That is one use case but it's not the only use case.

Could there be a better name here that conveys that these are values that were present . Does "values" not work ? (Although I can understand the reason not to just name the values)

Also shouldnt it use a record? Like we wouldn't want to show duplicate values right?

Finally I think we should limit the size of the values. If there is too high of a cardinality then it might be too much values to suggest. Perhaps this could be name like "recentValues" or "latestEntries"  because if there is like a 100 different unique values we can probably assume we are getting the values it whatever the user preferred. 

I can see a good argument that the user would only would want to see or need to see up to 10 unique records values. For example if a user's service is experiencing outages.  And most of their logs are 400s 404s 500 etc then probably they won't care to see as much 200s in their auto complete. They likely just want to see the error fast

kavilla commented on src/plugins/data/public/antlr/shared/utils.test.ts:
```diff
@@ -20,6 +20,13 @@ import { HttpSetup } from 'opensearch-dashboards/public';
 
+// Mock the getDataViews service
+jest.mock('../../services', () => ({
+  getDataViews: jest.fn(() => ({
```
Do we have a getdataviewsmockcontract if not I will make one

kavilla commented on src/plugins/data/common/index_patterns/types.ts:
```diff
@@ -185,6 +185,8 @@ export interface FieldSpec {
   subType?: IFieldSubType;
   indexed?: boolean;
+  autoCompleteValues?: string[];
+  topQueryValues?: string[];
```
What if instead of two new fields we make one new field that is a object that has these two values set on them. The reason why is because I feel like these values should be in sync with each other and updated together

kavilla commented on src/plugins/data/public/antlr/shared/utils.ts:
```diff
@@ -104,49 +107,94 @@ export const fetchColumnValues = async (
+    }
+
+    const limit = services.uiSettings.get(UI_SETTINGS.QUERY_ENHANCEMENTS_SUGGEST_VALUES_LIMIT);
+    const dataset = services.data.query.queryString.getQuery().dataset;
```
I think you should just use the index pattern passed in . The best option tho is if you have passed in the query object. At which point u access query.dataset to get the type of dataset and the full index pattern object from the service. Like directly passed in not from accessing a Singleton. To protect ourselves from race conditions

kavilla commented on src/plugins/data/public/antlr/shared/utils.ts:
```diff
@@ -104,49 +107,94 @@ export const fetchColumnValues = async (
+    const dataset = services.data.query.queryString.getQuery().dataset;
+
+    const values = (
+      await fetchFromAPI(
```
Oh we don't just use the SQL search strategy via search.search ? We should prolly do that vs making raw http calls if we are

kavilla commented on src/plugins/explore/public/components/fields_selector/lib/get_details.ts:
```diff
@@ -68,5 +69,12 @@ export function getDetails(
   }
+
+  // Save updated index pattern with topQueryValues to cache
+  if (indexPattern && indexPattern.id) {
```
If you could just update to dataView. Or dataset since that is where we want to be

kavilla commented on src/plugins/explore/public/components/fields_selector/lib/get_details.ts:
```diff
@@ -68,5 +69,12 @@ export function getDetails(
+
+  // Save updated index pattern with topQueryValues to cache
+  if (indexPattern && indexPattern.id) {
+    const services = getServices();
```
Is this from the context or could. We just use the context ?

kavilla commented on src/plugins/explore/public/components/fields_selector/lib/get_details.ts:
```diff
@@ -33,6 +33,7 @@ import { i18n } from '@osd/i18n';
+import { getServices } from '../../../application/legacy/discover/opensearch_dashboards_services';
 
 export function getDetails(
   field: IndexPatternField,
```
Update to use the right type for me? Then we can modify that file instead of touching. Index pattern field

kavilla reviewed (commented):
Could we mark these fields with the experimental annotation ?  I think we should really consider adding an explicit Boolean or specifying in the limits UI setting . That setting to -1 will disable field value caching. Just in case we need to protect ourselves and give the users a backdoor If we found out we might have missed something or if we cause blot.


Also I think we should have more explicit names or types. Could u also include an example values (, could be online comments )

Thank you for this. Cool stuff

LDrago27 commented:
> Could we mark these fields with the experimental annotation ? I think we should really consider adding an explicit Boolean or specifying in the limits UI setting . That setting to -1 will disable field value caching. Just in case we need to protect ourselves and give the users a backdoor If we found out we might have missed something or if we cause blot.
> 
> Also I think we should have more explicit names or types. Could u also include an example values (, could be online comments )
> 
> Thank you for this. Cool stuff

Hi @kavilla , Thanks for comments. Based on our offline sync. These are some of the things I have updated.
- Use Search Source for making the call 
- Updated the structure of newly added fields
- Updated the API to fetch column Values
- Moved out the allowed datatypes for suggestion to IndexPatternFields that now exposes a method isSuggestionAvailable to  provide this information

kavilla commented on src/plugins/data/common/search/search_source/types.ts:
```diff
@@ -104,6 +104,7 @@ export interface SearchSourceFields {
   timeout?: string;
   terminate_after?: number;
   df?: IDataFrame;
+  skipTimeFilter?: boolean;
```
whats this for?

LDrago27 replied:
Currently the PPL Interceptor always appends the timeFilter in the query. 

SQL language doesn't have a timePicker, but uses this util to fetch column Values. Since the PPL interceptor always adds in a timefilter so there are times SQL timepicker is set to default of 15 mins which results in us not receiving the appropriate results. In a bid to maintain backwards compatibility with SQL language, which don't have a timePicker, I am adding this as a way to bypass adding the timeFilter.

kavilla commented on src/plugins/data/common/index_patterns/types.ts:
```diff
@@ -185,6 +185,15 @@ export interface FieldSpec {
+   * TODO: Refactor and move them into DataViewField
+   */
+  suggestions?: {
+    autoCompleteValues?: string[];
```
nit: one more pitch if this could be just `values`. autocomplete is one user but it doesn't have to be the only user of this new field

kavilla commented on src/plugins/data/common/index_patterns/types.ts:
```diff
@@ -185,6 +185,15 @@ export interface FieldSpec {
+   */
+  suggestions?: {
+    autoCompleteValues?: string[];
+    topAggValues?: string[];
```
if we are skipping the timeFilter and if we are building for observability which usually involve time. should we just always be using the time filter then.

then we could just call it `topValues` which would give us a clean interface like

```
suggestions: {
   values: []
   topValues: []
}

which would make this reusuable

LDrago27 replied:
Sounds good.

LDrago27 commented on src/plugins/data/public/antlr/shared/utils.ts:
```diff
@@ -104,49 +107,94 @@ export const fetchColumnValues = async (
+    }
+    if (
+      !services.uiSettings.get(UI_SETTINGS.QUERY_ENHANCEMENTS_SUGGEST_VALUES) ||
+      !fieldInOsd ||
```
Actually the main reason to allow for undefined is for typescript.

LDrago27 commented on src/plugins/data/public/antlr/shared/utils.ts:
```diff
@@ -104,49 +107,94 @@ export const fetchColumnValues = async (
+    }
+  } catch (error) {
+    // Propagate the error instead of silently failing
+    throw error;
```
Agreed. We werenot doing anything. Currently removed it to allow for it fail silently without blocking the cx.

LDrago27 commented:
@kavilla I liked your idea of having the suggestion being filtered by time. 
So instead of removing the skipTimeFilter param that I introduced I felt it is better to have controls. 

For all queries we will be setting the param as False except in SQL where we set is as True. Since we have the limitation of SQL not having timePicker which actually sets the time range in the search source.

sejli commented on src/plugins/explore/public/application/utils/state_management/actions/query_actions.ts:
```diff
@@ -91,13 +96,69 @@ export const defaultResultsProcessor: DefaultDataProcessor = (
+        fieldUpdates.push({ field, topValues });
+      }
+    } catch (error) {
+      // Silently continue on field processing errors
```
Do we want to log this somewhere and continue?

LDrago27 commented on src/plugins/explore/public/application/utils/state_management/actions/query_actions.ts:
```diff
@@ -91,13 +96,69 @@ export const defaultResultsProcessor: DefaultDataProcessor = (
+      // Update the IndexPattern field
+      const indexPatternField = dataset.fields.getByName(field.name);
+      if (indexPatternField) {
+        const indexPatternFieldWithSuggestions = indexPatternField as any;
```
@sejli Thanks. Removed the any.

sejli commented on src/plugins/data/common/index_patterns/fields/index_pattern_field.ts:
```diff
@@ -148,6 +148,10 @@ export class IndexPatternField implements IFieldType {
     return this.aggregatable && !notVisualizableFieldTypes.includes(this.spec.type);
   }
 
+  public isSuggestionAvailable() {
```
Nit: May want to consider renaming this function. Initially, it looked like it was to check if there's a suggestion available, when its actual purpose is to check if it's possible to generate suggestions on this field. For example, like `isSuggestable` or something of the sort.

LDrago27 replied:
Thanks. Will clean up in a followup PR
